### ISSUE TITLE
Some questions about dataloading and model

### ISSUE THREAD

**Issue by L4zyy** [decisionforce/mmTransformer#5]
Hi, congrats on the nice work and thank you for the quick replies on other issues and sharing the data preprocessing repo, which really helps me a lot. I have some questions about the VectorNet and training.

(a) I wonder if you use the subgraph implemented in [this](https://github.com/xk-huang/yet-another-vectornet) repo? I am new to GNN and torch_geometric, I wonder if I can just implement the model with both torch and torch_geometric?

(b) How many epochs do you train for a single experiment? My implementation can not get a good results(2.0+ minADE and 5.0+ minFDE) and I found it takes much more epochs for my model to overfit a small subset of data compare to other non-transformer based model. I wonder if my implementation has some bugs or my training process is wrong.

**Comment by Mrmoore98** [comment]
> Hi, congrats on the nice work and thank you for the quick replies on other issues and sharing the data preprocessing repo, which really helps me a lot. I have some questions about the VectorNet and training.
> 
> (a) I wonder if you use the subgraph implemented in [this](https://github.com/xk-huang/yet-another-vectornet) repo? I am new to GNN and torch_geometric, I wonder if I can just implement the model with both torch and torch_geometric?

Yes, we follow that repo's subgraph implementation.
 Using our preprocess code you only need torch, and torch geometric is not necessary. 

> (b) How many epochs do you train for a single experiment? My implementation can not get a good results(2.0+ minADE and 5.0+ minFDE).

Model will converge quickly(within 10 epoch you may get the result that MR<15% ) and takes 60 epochs to reach the best performance. 

>  I found it takes much more epochs for my model to overfit a small subset of data compare to other non-transformer based model. 

The reason is that transformer is more complicated than other model like LSTM and the regurization of Transformer is weaker. For the small training set you may consider to scale down the model size.

> I wonder if my implementation has some bugs or my training process is wrong.

The model is not convergent in your case. So the most possible reason may lies in the data preprocess stage. I recommand you that fisrt reproducing the simplest form of mmTransformer, which only have motion extractor and 6proposals, to check your other parts are all bug-free.

Other suggestions:
1. Check the loss, whether it only update the best performance trajectory.
2. Check whether you have rotated the target car or not.
3. Check whether the preprocess data are lossless or not.

**Comment by L4zyy** [comment]
Thank you again for the quick reply and all these helpful suggestions!!! I will check these things and reimplement my model.

**Comment by MohammadHossein-Bahari** [comment]
Hi, 
I'm trying to connect the scene processing part to my Transformers. I just wonder how to use the two repositories. You mentioned [this repo](https://github.com/jzhang538/Argoverse-Data-Preprocess) for preprocessing and [this repo](https://github.com/xk-huang/yet-another-vectornet) for vectornet implementation. However, they do not match in some details, like the way they make dictionaries. Can you help with this?

Thanks!

**Comment by L4zyy** [comment]
Hi, with your help, I finally figured out that the bug is in my rotation code. After modify my rotation code I can get a reasonable result with only motion extractor. However, I still have a problem similar to @MohammadHossein-Bahari . 

> Yes, we follow that repo's subgraph implementation.
Using our preprocess code you only need torch, and torch geometric is not necessary.

In [this repo](https://github.com/xk-huang/yet-another-vectornet), they mainly use torch geometric in the data part and the subgraph's max_pool function and GraphLayerProp layer. I wonder if you change these modules in subgraph to bypass the torch geometric? If so, how do you implement the propogate and maxpool functions? If not, how do you load the data preprocessed by [this repo](https://github.com/jzhang538/Argoverse-Data-Preprocess) to feed to the subgraph module? Thank you!

**Comment by jzhang538** [comment]
> Hi, with your help, I finally figured out that the bug is in my rotation code. After modify my rotation code I can get a reasonable result with only motion extractor. However, I still have a problem similar to @MohammadHossein-Bahari .
> 
> > Yes, we follow that repo's subgraph implementation.
> > Using our preprocess code you only need torch, and torch geometric is not necessary.
> 
> In [this repo](https://github.com/xk-huang/yet-another-vectornet), they mainly use torch geometric in the data part and the subgraph's max_pool function and GraphLayerProp layer. I wonder if you change these modules in subgraph to bypass the torch geometric? If so, how do you implement the propogate and maxpool functions? If not, how do you load the data preprocessed by [this repo](https://github.com/jzhang538/Argoverse-Data-Preprocess) to feed to the subgraph module? Thank you!

It is a good question. Actually, we calculate the maximum number of lanes and neighbors in all the cases and pad 0 to get fixed input tensors (vectorized lanes and neighbors) for each scenario. In this way, we do not need index-related functions in torch geometric to do calculations. With such implementation, the subgraph module can be represented as a series of Linear and Maxpooling functions that are available in torch.nn library.

We will make the Dataset.py with detailed implementation publicly available very soon.

**Comment by L4zyy** [comment]
Thank you again for the quick and clear reply. I will close this issue and start a new one if I have other questions.
